tools/bcr_validation.py

```diff
@@ -164,18 +170,78 @@ def verify_presubmit_yml_change(registry, module_name, version):
-  validation_results.extend(verify_presubmit_yml_change(registry, module_name, version))
+  source = registry.get_source(module_name, version)
+  source_url = source["url"]
+  tmp_dir = Path(tempfile.mkdtemp())
+  archive_file = tmp_dir.joinpath(source_url.split("/")[-1])
+  output_dir = tmp_dir.joinpath("source_root")
+  download_file(source_url, archive_file)
+  shutil.unpack_archive(str(archive_file), output_dir)
+
+  # Apply patch files if there are any, also verify their integrity values
+  source_root = output_dir.joinpath(source["strip_prefix"] if "strip_prefix" in source else "")
+  if "patches" in source:
+      for patch_name, expected_integrity in source["patches"].items():
+          patch_file = registry.get_patch_file_path(module_name, version, patch_name)
+          actual_integrity = integrity(read(patch_file))
+          if actual_integrity != expected_integrity:
+            validation_results.append((BcrValidationResult.FAILED, f"The patch file `{patch_file}` has expected integrity value `{expected_integrity}`, but the real integrity value is `{actual_integrity}`."))
+          apply_patch(source_root, source["patch_strip"], str(patch_file.resolve()))
+
+  source_module_dot_bazel = source_root.joinpath("MODULE.bazel")
+  if source_module_dot_bazel.exists():
+    source_module_dot_bazel_content = open(source_module_dot_bazel, "r").readlines()
+  else:
+    source_module_dot_bazel_content = []
+  bcr_module_dot_bazel_content = open(registry.get_module_dot_bazel_path(module_name, version), "r").readlines()
+  source_module_dot_bazel_content = remove_trailing_empty_lines(source_module_dot_bazel_content)
+  bcr_module_dot_bazel_content = remove_trailing_empty_lines(bcr_module_dot_bazel_content)
+  file_name = "a/" * int(source.get("patch_strip", 0)) + "MODULE.bazel"
+  diff = list(unified_diff(source_module_dot_bazel_content, bcr_module_dot_bazel_content, fromfile=file_name, tofile=file_name))
+
+  if diff:
+    validation_results.append((BcrValidationResult.FAILED,
+                               "Checked in MODULE.bazel file doesn't match the one in the extracted and patched sources.\n"
+                               + f"Please fix the MODULE.bazel file or you can add the following patch to {module_name}@{version}:\n"
+                               + "    " + "    ".join(diff)))
+  else:
+    validation_results.append((BcrValidationResult.GOOD, "Checked in MODULE.bazel matches the sources."))
+
+  shutil.rmtree(tmp_dir)
 
-  print_validation_result(validation_results)
   return validation_results
 
+def print_and_append_result(all_results, new_results):
+  print_validation_result(new_results)
+  all_results.extend(new_results)
+
+def validate_module(registry, module_name, version):
+  print_expanded_group(f"Validating {module_name}@{version}")
+
+  all_results = []
+  print_and_append_result(all_results, verify_module_existence(registry, module_name, version))
```

Comment:
is this change so that we still print earlier verification results if later ones raise an exception?

if so, I wonder if it'd be easier if we make a class and store the results in there, make all `verify_*` functions its methods, and instead of appending a pair to `validation_results` every time, just have a method `self.report(BcrValidationResult.X, "message")`. That would reduce some code noise IMO

Reply:
Good idea, done!